### ISSUE TITLE
Add ability to pass compute buffers to vertex/fragment shaders

### ISSUE THREAD

**Issue by jtsorlinis** [godotengine/godot-proposals#6989]
### Describe the project you are working on

Procedural mesh generation in a compute shaders.

### Describe the problem or limitation you are having in your project

The buffers containing the mesh data have to be read back and processed on the CPU into vertex buffers.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

I think it would be really helpful to allow storage buffers to be bound to and also accessed in vertex/fragment shaders.

This would greatly expand the potential uses for compute shaders as the current implementation requires CPU readback which can be a significant performance bottleneck.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

A storage buffer can be bound to a vertex/fragment shader with a function similar to  set_shader_parameter eg
``` 
var buffer := rd.storage_buffer_create(input_bytes.size(), input_bytes)
myShaderMaterial.set_shader_parameter("myBuffer", buffer) # something like this
```

Which could then be used in said shader to allow procedural drawing eg.
```
shader_type spatial;

layout(set = 0, binding = 0, std430) restrict buffer bufferType {
    vec4 positions[];
} myBuffer;

void vertex() {
  VERTEX += myBuffer.positions[VERTEX_ID];
}

void fragment() {
  COLOR = vec4(0.4, 0.6, 0.9, 1.0);
}
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I don't believe it's possible to implement this as a plugin as its part of the core rendering pipeline.

### Is there a reason why this should be core and not an add-on in the asset library?

As above, I don't think it's possible.

**Comment by NomiChirps** [comment]
+1! I'm working on a project that requires simulating an arbitrarily deformable/cuttable/etc object with high precision. I'm currently using a 512x512x512 3D texture to store an SDF representing the object's surface. Writing a fragment shader to render it was "no problem" (ha ha), but making changes to it from the CPU is extremely slow. I'd love it if Godot could make it easy to use a compute shader to directly modify such a buffer on the GPU.

**Comment by Daedalus1400** [comment]
Let's extend this to it's logical conclusion: All shaders should be able to access compute buffers, and compute shaders should be able to access render buffers.

I'm currently working on a compute shader based particle simulation, and the frame rate is terrible for large simulations despite neither my CPU nor GPU being taxed.  It's bottle necked by writing the position data to a MultiMeshInstance3D from the CPU.  If I could access the position buffer from inside a particle shader, the particles would position themselves with no CPU overhead.

The complete memory isolation of gdshader scripts is a huge limitation.

**Comment by wojtekpil** [comment]
I totally agree on importance of this proposal. Without it the usages of compute shaders are very limited. With possibility of editing buffers and textures from compute shaders we create opportunity for custom GPU based culling systems, custom LOD systems (even for multimeshes - imagine that you could use LOD per instance). Add to it possibility of reading render buffers from CS  and even post processing effects that are hard to achieve otherwise could be easily added and chained (e.g. gaussian blur for some effects needs 2-3 viewports, that could be chained in 2 simple compute shaders). It could also help to avoid issues with transparency of full screen post-processing planes in front of camera. Also it would probably make life easier for creating things like GPU painting applications or terrain editors, as we are not limited by one 8-bit viewport texture.

**Comment by Facundo15** [comment]
It must definitely be something integrated into Godot, I am working on a projectile simulation system in which I am calculating the movement, and a very primitive physics (rectangles and buffer of all rectangular shapes) to detect collisions and send them to the CPU.
But I use Particles to visualize the behavior and having to send everything to the CPU for then the particle shader became a very slow process.

**Comment by Facundo15** [comment]
@jtsorlinis I have a better way that can be implemented at the code level when you want to make that association.

Basically it is that the buffers are "shared" so that the Compute Shader can directly write uniforms as it wants within the godot shader and it can be something like

```
var buffer := rd.storage_buffer_create(input_bytes.size(), input_bytes)
material_shader.set_shared_uniform("unform_data", buffer_rid);

```

```
shader_type canvas_item;

shared uniform vec2 uniform_data;
```

This might be a simpler way to tell the shader code that the buffer will be shared between the compute shader and the godot shader (a simpler form of adding fo following Godot's simple philosophy).

I'm not an expert in glsl and shaders, so I don't know and the shared keyword could be used for this case.

**Comment by oxi-dev0** [comment]
I also agree with this proposal. I am currently working on a surface fluid system to create effects like blood running down a character, and in order to render the calculated trail to a mask i opted to use a compute shader so i have total control over the final trail width, colour, roll-off, etc. I am new to Godot, and may be misunderstanding the datatypes, but I think this proposal should also mention that it would be really beneficial for compute shaders and materials to be able to directly share ImageTexture resources, as well as storage buffers.

**Comment by BenMcLean** [comment]
Because of this issue, I invested a good deal of time working out how to treat a texture uniform as if it was a storage buffer to "smuggle" my raw byte data onto the GPU as a workaround.
https://gist.github.com/BenMcLean/9327690b93690b8a92a921df003f7954

**Comment by TokisanGames** [comment]
@BenMcLean You can do this [more directly](https://github.com/godotengine/godot/issues/57841#issuecomment-1773930802), reading and writing 32-bit uints on the CPU and in the shader. 32-bit floats are even easier. There's no need to use an rgba8 format and having to convert numbers and risk precision loss. 

For a working example of transferring non-image data to the shader in production, we transfer a texture array full of up to 1GB (16k^2 32-bit) of bit packed uint data. You can look through the code for [Terrain3D](https://github.com/TokisanGames/Terrain3D), our bit packed [control map format](https://terrain3d.readthedocs.io/en/stable/docs/controlmap_format.html), [en/decoders](https://github.com/TokisanGames/Terrain3D/blob/1ff769d816608410e3292112d6d74bcd687c5c3d/src/util.h#L17), [CPU writer](https://github.com/TokisanGames/Terrain3D/blob/main/src/terrain_3d_editor.cpp#L253-L326), [shader](https://github.com/TokisanGames/Terrain3D/blob/1ff769d816608410e3292112d6d74bcd687c5c3d/src/shaders/main.glsl#L30) and [here](https://github.com/TokisanGames/Terrain3D/blob/1ff769d816608410e3292112d6d74bcd687c5c3d/src/shaders/auto_shader.glsl#L23).

------

**Regarding this ticket**
compute -> vertex/fragment without copying to the CPU, doesn't [Bastian's water compute demo](https://github.com/godotengine/godot-demo-projects/tree/master/compute/texture) already do this using a `Texture2DRD`? He says `Instead of copying data from texture to texture to create this history, we simply cycle the RIDs.` and in the code I don't see any functions that pull the Texture from compute to an Image. It just gets the texture RID from the compute shader and assigns it to the shader material uniform.

**Comment by kb173** [comment]
> **Regarding this ticket** compute -> vertex/fragment without copying to the CPU, doesn't [Bastian's water compute demo](https://github.com/godotengine/godot-demo-projects/tree/master/compute/texture) already do this using a `Texture2DRD`? He says `Instead of copying data from texture to texture to create this history, we simply cycle the RIDs.` and in the code I don't see any functions that pull the Texture from compute to an Image. It just gets the texture RID from the compute shader and assigns it to the shader material uniform.

This issue (as I understand it) is about passing vertex data, i.e. meshes, directly from compute to vertex shader to facilitate procedural mesh generation, right? So workarounds with textures don't work because we still can't create new vertices without involving the CPU.

**Comment by clayjohn** [comment]
Just to add some context here. We already have plans to expose a way to create/update meshes using compute shaders without copying the data to the CPU and back https://github.com/godotengine/godot-proposals/issues/7209

Using individual SSBO's for each channel will be a bit too cumbersome I think and will be pretty limiting as it wouldn't allow you to create meshes totally on the GPU and then use them in rendering.

**Comment by xRoGeSx** [comment]
Hey, Idk if it's too late, but I was able to acive that by switching from 
`RenderingServer.get_rendering_device()`
to
`RenderingServer.create_local_rendering_device()`

Rest is simple:
- Create texture buffer from desired image data
```
boid_data_buffer = rd.texture_create(fmt, view,  boid_data.get_data())
boid_data_texture_rd.texture_rd_rid = boid_data_buffer
boid_data_uniform = createUniformFromBuffer(boid_data_buffer, 4, 0, RenderingDevice.UNIFORM_TYPE_IMAGE)
```
- Get sampler2D texture from your Particle shader
`boid_data_texture_rd = $BoidParticle.process_material.get_shader_parameter("boid_data")`
- Replace buffer reference with your own buffer
`boid_data_texture_rd.texture_rd_rid = boid_data_buffer`


Now you have no CPU bottleneck and both shaders are sharing a buffer

**Comment by cheesebellies** [comment]
I need this to access custom per-vertex data in a vertex shader. Even without full support, there should at least be a way to bind an SSBO to a shader (`RenderingServer.material_bind_buffer(material, buffer)` or similar.

**Comment by tracefree** [comment]
@cheesebellies You can actually already add custom vertex data. If using an ArrayMesh, you can add data to one of the [ARRAY_FORMAT_CUSTOM0/1/2/3](https://docs.godotengine.org/en/stable/classes/class_mesh.html#enum-mesh-arrayformat) array types, and access it in the vertex shader with the built-in [CUSTOM0/1/2/3](https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/spatial_shader.html#vertex-built-ins).

**Comment by cmharris** [comment]
> [@cheesebellies](https://github.com/cheesebellies) You can actually already add custom vertex data. If using an ArrayMesh, you can add data to one of the [ARRAY_FORMAT_CUSTOM0/1/2/3](https://docs.godotengine.org/en/stable/classes/class_mesh.html#enum-mesh-arrayformat) array types, and access it in the vertex shader with the built-in [CUSTOM0/1/2/3](https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/spatial_shader.html#vertex-built-ins).

ArrayMesh requires CPU readback. This proposal is to enable mesh creation on GPU without having to go through the bottleneck that is the GPU -> CPU bus.

**Comment by tracefree** [comment]
@cmharris Sure, I just mentioned an alternative way to use custom vertex data that already exists in case the use case of the person I replied to doesn't require generating that data on the GPU.
